`basilisp test` fails when attempting to `print` anything out from inside a test file on MS-Windows

Hi,

when running `basilisp test`, an error occurs when trying to `print`  from a test file
```
    (.write stdout os/linesep)
E   OSError: [WinError 6] The handle is invalid
```

To reproduce, 
1. Clone https://github.com/ikappaki/issue-bas-test-macro-print example project, and setup 
``` shell
git clone https://github.com/ikappaki/issue-bas-test-macro-print
cd issue-bas-test-macro-print
poetry install
poetry shell
```
2. Run the tests with `basilisp`, you should see the error 
``` shell
basilisp test

==================================================== test session starts ====================================================
platform win32 -- Python 3.11.4, pytest-8.3.3, pluggy-1.5.0
rootdir: D:\bas\issue-bas-test-macro-print
configfile: pyproject.toml
plugins: basilisp-0.2.4
collected 0 items / 1 error

========================================================== ERRORS ===========================================================
_____________________________________ ERROR collecting tests/issuetests/issue_test.lpy ______________________________________
...
During handling of the above exception, another exception occurred:
tests\issuetests\issue_test.lpy:4: in <module>
    (println :calc (+ 5 3))
..\issuetests\.venv\Lib\site-packages\basilisp\core.lpy:4376: in println
    (defn println
..\issuetests\.venv\Lib\site-packages\basilisp\core.lpy:4398: in println__arity_rest
    (.write stdout os/linesep)
E   OSError: [WinError 6] The handle is invalid
===================================================== warnings summary ======================================================
..\issuetests\.venv\Lib\site-packages\_pytest\config\__init__.py:1277
  d:\bas\issuetests\.venv\Lib\site-packages\_pytest\config\__init__.py:1277: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: basilisp
    self._mark_plugins_for_rewrite(hook)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================== short test summary info ==================================================
ERROR tests/issuetests/issue_test.lpy - OSError: [WinError 6] The handle is invalid
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================ 1 warning, 1 error in 0.11s ================================================
:calc 8
```

The test file just tries to printout a value (I've also added a test event though is irrelevant)
``` clojure
(ns tests.issuetests.issue-test
  (:require [basilisp.test :refer [deftest are is testing]]))

(println :calc (+ 5 3))

(deftest 5-test []
  (is (= 5 5)))
```

Commenting out the `println` form allows the test to run without error.

This could as well be a MS-Windows only issue, I haven't had the chance to test on another architecture yet.

Thanks

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Fixed
+ * Fix an issue with `basilisp test` standard streams output that can lead to failures on MS-Windows (#1080)
 
 ## [v0.2.4]
 ### Added

--- a/src/basilisp/contrib/pytest/testrunner.py
+++ b/src/basilisp/contrib/pytest/testrunner.py
@@ -42,21 +42,19 @@ def pytest_configure(config):
     # during tests and restore them afterward.
     out_var = runtime.Var.find(OUT_VAR_SYM)
     err_var = runtime.Var.find(ERR_VAR_SYM)
-    if out_var:
-        out_var.push_bindings(sys.stdout)
-        config.out_var = out_var
-    if err_var:
-        err_var.push_bindings(sys.stderr)
-        config.err_var = err_var
+    bindings = {
+        k: v for k, v in {out_var: sys.stdout, err_var: sys.stderr}.items() if k
+    }
+    if bindings.items():
+        runtime.push_thread_bindings(bindings)
+        config.basilisp_bindings = bindings
 
     basilisp.bootstrap("basilisp.test")
 
 
 def pytest_unconfigure(config):
-    if hasattr(config, "out_var"):
-        config.out_var.pop_bindings()
-    if hasattr(config, "err_var"):
-        config.err_var.pop_bindings()
+    if hasattr(config, "basilisp_bindings"):
+        runtime.pop_thread_bindings()
 
 
 def pytest_collect_file(  # pylint: disable=unused-argument

--- a/src/basilisp/contrib/pytest/testrunner.py
+++ b/src/basilisp/contrib/pytest/testrunner.py
@@ -46,7 +46,7 @@ def pytest_configure(config):
         k: v for k, v in {out_var: sys.stdout, err_var: sys.stderr}.items() if k
     }
     if bindings.items():
-        runtime.push_thread_bindings(bindings)
+        runtime.push_thread_bindings(lmap.map(bindings))
         config.basilisp_bindings = bindings
 
     basilisp.bootstrap("basilisp.test")

--- a/src/basilisp/contrib/pytest/testrunner.py
+++ b/src/basilisp/contrib/pytest/testrunner.py
@@ -1,6 +1,7 @@
 import importlib.util
 import inspect
 import os
+import sys
 import traceback
 from pathlib import Path
 from types import GeneratorType
@@ -12,6 +13,7 @@ from basilisp import main as basilisp
 from basilisp.lang import keyword as kw
 from basilisp.lang import map as lmap
 from basilisp.lang import runtime as runtime
+from basilisp.lang import symbol as sym
 from basilisp.lang import vector as vec
 from basilisp.lang.obj import lrepr
 from basilisp.util import Maybe
@@ -20,13 +22,46 @@ _EACH_FIXTURES_META_KW = kw.keyword("each-fixtures", "basilisp.test")
 _ONCE_FIXTURES_NUM_META_KW = kw.keyword("once-fixtures", "basilisp.test")
 _TEST_META_KW = kw.keyword("test", "basilisp.test")
 
+CORE_NS = "basilisp.core"
+CORE_NS_SYM = sym.symbol(CORE_NS)
+OUT_VAR_NAME = "*out*"
+OUT_VAR_SYM = sym.symbol(OUT_VAR_NAME, ns=CORE_NS)
+ERR_VAR_NAME = "*err*"
+ERR_VAR_SYM = sym.symbol(ERR_VAR_NAME, ns=CORE_NS)
 
-# pylint: disable=unused-argument
+
 def pytest_configure(config):
+
+    # https://github.com/pytest-dev/pytest/issues/12876
+    #
+    # Basilisp's standard output streams may be initialized before
+    # pytest captures sys streams (sys.stdout and sys.stderr) for
+    # testing (e.g., with `basilisp test`). Writing to the original
+    # handles during tests on Windows can cause invalid handle
+    # errors. To prevent this, we rebind them to pytest's streams
+    # during tests and restore them afterward.
+    out_var = runtime.Var.find(OUT_VAR_SYM)
+    err_var = runtime.Var.find(ERR_VAR_SYM)
+    if out_var:
+        out_var.push_bindings(sys.stdout)
+        config.out_var = out_var
+    if err_var:
+        err_var.push_bindings(sys.stderr)
+        config.err_var = err_var
+
     basilisp.bootstrap("basilisp.test")
 
 
-def pytest_collect_file(file_path: Path, path, parent):
+def pytest_unconfigure(config):
+    if hasattr(config, "out_var"):
+        config.out_var.pop_bindings()
+    if hasattr(config, "err_var"):
+        config.err_var.pop_bindings()
+
+
+def pytest_collect_file(  # pylint: disable=unused-argument
+    file_path: Path, path, parent
+):
     """Primary PyTest hook to identify Basilisp test files."""
     if file_path.suffix == ".lpy":
         if file_path.name.startswith("test_") or file_path.stem.endswith("_test"):

